### PR TITLE
Added list.size() instruction to IESI

### PR DESCRIPTION
**Pull request by Suyash-d**
Hi Robbe,
Please review changes for newly added ListSize instruction

### CODE REVIEW

File: core/java/iesi-core/src/main/java/io/metadew/iesi/script/execution/ExecutionRuntime.java [hunk 1]
```diff
@@ -85,7 +85,7 @@ public ExecutionRuntime(ExecutionControl executionControl, String runId) {
         // Initialize extensions
 
         // Initialize data instructions
-        dataInstructions = DataInstructionRepository.getRepository(new GenerationObjectExecution());
+        dataInstructions = DataInstructionRepository.getRepository(new GenerationObjectExecution(),this);
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/script/execution/ExecutionRuntime.java:88]: code formatting. you can auto-format the code using IntelliJ.
Normally the Sonarlint plugin will also complain about this.

File: core/java/iesi-core/src/main/java/io/metadew/iesi/script/execution/instruction/data/DataInstructionRepository.java [hunk 2]
```diff
@@ -57,6 +55,9 @@
         TextSubstring textSubstring = new TextSubstring();
         dataInstructions.put(textSubstring.getKeyword(), textSubstring);
 
+        ListSize listSize = new ListSize(executionRuntime);
+        dataInstructions.put(listSize.getKeyword(),listSize);
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/script/execution/instruction/data/DataInstructionRepository.java:59]: code formatting. you can auto-format the code using IntelliJ.
Normally the Sonarlint plugin will also complain about this.

File: core/java/iesi-core/src/main/java/io/metadew/iesi/script/execution/instruction/data/date/ListSize.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        List<DataType> list = array.getList();
+        return Integer.toString(list.size());
+    }
+
+    private Array getArray(DataType array) {
+        if (array instanceof Array) {
+            return (Array) array;
+        } else if (array instanceof Text) {
+            return executionRuntime.getArray(((Text) array).getString())
+                    .orElseThrow(() -> new IllegalArgumentException(MessageFormat.format("No array found", ((Text) array).getString())));
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/main/java/io/metadew/iesi/script/execution/instruction/data/date/ListSize.java:40]: replace `MessageFormat.format("No array found", ((Text) array).getString()))` with `String.format("No array %s found in memory", ((Text) array).getString()))`

File: core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        ListSize listSize = new ListSize(executionRuntime);
+        assertThat(listSize.getKeyword()).isEqualTo("list.size");
+    }
+
+    @Test
+    public void generateOutputWithText() {
+        DataTypeHandler dataTypeHandler = DataTypeHandler.getInstance();
+        DataTypeHandler dataTypeHandlerServiceSpy = Mockito.spy(dataTypeHandler);
+        Whitebox.setInternalState(DataTypeHandler.class, "instance", dataTypeHandlerServiceSpy);
+        ExecutionRuntime executionRuntime = mock(ExecutionRuntime.class);
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java:35]: Have a look online on @Before and @After for unit tests. This recurring block can be put in setup and teardown methods for the tests

File: core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        doReturn(new Text("array")).when(dataTypeHandlerServiceSpy).resolve("array", executionRuntime);
+        when(executionRuntime.getArray("array")).thenReturn(Optional.of(new Array(new ArrayList<>())));
+        ListSize listSize = new ListSize(executionRuntime);
+        assertThat(listSize.generateOutput("array")).isEqualTo("0");
+        Whitebox.setInternalState(DataTypeHandler.class, "instance", (DataTypeHandler) null);
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java:40]: Have a look online on @Before and @After for unit tests. This recurring block can be put in setup and teardown methods for the tests

File: core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Test
+    public void generateOutputWithNull() {
+        DataTypeHandler dataTypeHandler = DataTypeHandler.getInstance();
+        DataTypeHandler dataTypeHandlerServiceSpy = Mockito.spy(dataTypeHandler);
+        Whitebox.setInternalState(DataTypeHandler.class, "instance", dataTypeHandlerServiceSpy);
+        ExecutionRuntime executionRuntime = mock(ExecutionRuntime.class);
+        doReturn(new Null()).when(dataTypeHandlerServiceSpy).resolve(null, executionRuntime);
+        ListSize listSize = new ListSize(executionRuntime);
+        assertThrows(IllegalArgumentException.class, () -> listSize.generateOutput(null));
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java:64]: Can you also check the Exception message?

File: core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Test
+    public void getKeyword() {
+        ExecutionRuntime executionRuntime = mock(ExecutionRuntime.class);
+        ListSize listSize = new ListSize(executionRuntime);
+        assertThat(listSize.getKeyword()).isEqualTo("list.size");
+    }
+
+    @Test
+    public void generateOutputWithText() {
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java:31]: Add a similar test where the array with name "array" is not found in memory (so ExecutionRuntime#getArray return an Optional.empty() )

**Review by BerrevoetsRobbe — CHANGES_REQUESTED** [review]

File: core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import java.util.ArrayList;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.*;
+
+public class ListSizeTest {
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java:21]: code formatting. 
public can be removed, normally the SonarLint plugin complains about this

File: core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Test
+    public void getKeyword() {
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java:24]: code formatting. 
public can be removed, normally the SonarLint plugin complains about this.

**Review by BerrevoetsRobbe — CHANGES_REQUESTED** [review]

**Review by Suyash-d — COMMENTED** [review]
Review Changes Added

File: core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java (hunk 6 above)
Comment (BerrevoetsRobbe) [reply core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java:79]: @Suyash-d reminder

File: core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java [hunk 10]
```diff
@@ -79,4 +79,23 @@ void generateOutputWithNull() {
         assertThrows(IllegalArgumentException.class, () -> listSize.generateOutput(null));
 
     }
+
+    @Test
+    void generateOutputWithExceptionMSG() {
+        doReturn(new Null()).when(dataTypeHandlerServiceSpy).resolve(null, executionRuntime);
+        listSize = new ListSize(executionRuntime);
+        Exception exception = assertThrows(IllegalArgumentException.class, () -> listSize.generateOutput(null));
+        assertThat(exception.getMessage()).isEqualTo("list cannot be of type class io.metadew.iesi.datatypes._null.Null");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java:88]: you can combine these tests using assertJ
https://www.baeldung.com/assertj-exception-assertion

File: core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java [hunk 11]
```diff
@@ -79,4 +79,23 @@ void generateOutputWithNull() {
...
+
+    }
+
+    @Test
+    void generateOutputWithOptionalArrayMessage() {
+        doReturn(new Text("array")).when(dataTypeHandlerServiceSpy).resolve("array", executionRuntime);
+        when(executionRuntime.getArray("array")).thenReturn(Optional.empty());
+        ListSize listSize = new ListSize(executionRuntime);
+        Exception exception = assertThrows(IllegalArgumentException.class, () -> listSize.generateOutput("array"));
+        assertThat(exception.getMessage()).isEqualTo("No array array found in memory");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java:98]: you can combine these tests using assertJ
https://www.baeldung.com/assertj-exception-assertion

File: core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java [hunk 12]
```diff
@@ -79,4 +79,23 @@ void generateOutputWithNull() {
...
+    }
+
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-core/src/test/java/io/metadew/iesi/script/execution/instruction/data/date/ListSizeTest.java:100]: on line 37: is the class variable `listSize` ever used?
If not, can you remove it?

**Review by BerrevoetsRobbe — CHANGES_REQUESTED** [review]

**Review by Suyash-d — COMMENTED** [review]
Requested Review Changes Added

**Review by BerrevoetsRobbe — APPROVED** [review]
approved

**Pull request merged**